src/common/procedure/src/local/runner.rs

```diff
@@ -53,6 +52,58 @@ impl ExecResult {
     }
 }
 
+/// A guard to cleanup procedure state.
+struct ProcedureGuard {
+    meta: ProcedureMetaRef,
+    manager_ctx: Arc<ManagerContext>,
+    finish: bool,
+}
+
+impl ProcedureGuard {
+    /// Returns a new [ProcedureGuard].
+    fn new(meta: ProcedureMetaRef, manager_ctx: Arc<ManagerContext>) -> ProcedureGuard {
+        ProcedureGuard {
+            meta,
+            manager_ctx,
+            finish: false,
+        }
+    }
+
+    /// The procedure is finished successfully.
+    fn finish(mut self) {
+        self.finish = true;
+    }
+}
+
+impl Drop for ProcedureGuard {
+    fn drop(&mut self) {
+        if !self.finish {
+            logging::error!("Procedure {} exits unexpectedly", self.meta.id);
+
+            // Set state to failed. This is useful in test as runtime may not abort when the runner task panics.
+            // See https://github.com/tokio-rs/tokio/issues/2002 .
+            // We set set_panic_hook() in the application's main function. But our tests don't have this panic hook.
+            self.meta.set_state(ProcedureState::Failed);
```

Comment:
Shall this unexpected drop happens, are the resources reclamations (notify, unlocks and cache invalidations) below still need to be executed? Or we can just return from here?

Reply:
We should always release locks and notify its parent procedure. Cache invalidation isn't necessary so I can only do it when the procedure is finished.